src/borg/testsuite/compress.py

```diff
@@ -221,51 +166,41 @@ def test_obfuscate_meta():
+    assert csize - psize >= 0  # there is an obfuscation trailer
     trailer = compressed[psize:]
     assert not trailer or set(trailer) == {0}  # trailer is all-zero-bytes
 
 
-def test_compression_specs():
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("")
-
-    assert isinstance(CompressionSpec("none").compressor, CNONE)
-    assert isinstance(CompressionSpec("lz4").compressor, LZ4)
-
-    zlib = CompressionSpec("zlib").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 6
-    zlib = CompressionSpec("zlib,0").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 0
-    zlib = CompressionSpec("zlib,9").compressor
-    assert isinstance(zlib, ZLIB)
-    assert zlib.level == 9
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("zlib,9,invalid")
-
-    lzma = CompressionSpec("lzma").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 6
-    lzma = CompressionSpec("lzma,0").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 0
-    lzma = CompressionSpec("lzma,9").compressor
-    assert isinstance(lzma, LZMA)
-    assert lzma.level == 9
-
-    zstd = CompressionSpec("zstd").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 3
-    zstd = CompressionSpec("zstd,1").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 1
-    zstd = CompressionSpec("zstd,22").compressor
-    assert isinstance(zstd, ZSTD)
-    assert zstd.level == 22
-
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("lzma,9,invalid")
-    with pytest.raises(argparse.ArgumentTypeError):
-        CompressionSpec("invalid")
+@pytest.mark.parametrize(
+    "c_type, c_name, c_levels",
+    [
+        (ZLIB, "zlib", ["", 0, 9, "9,invalid"]),
+        (LZMA, "lzma", ["", 0, 9, "9,invalid"]),
+        (ZSTD, "zstd", ["", 1, 22, "22,invalid"]),
+        (CNONE, "none", [""]),
+        (LZ4, "lz4", [""]),
+        (None, None, ["invalid"]),
+    ],
+)
+def test_compression_specs(c_type, c_name, c_levels):
+    for level in c_levels:
+        # a compression level is specified
+        if isinstance(level, int):
+            cs = CompressionSpec(f"{c_name},{level}").compressor
+            assert isinstance(cs, c_type)
+            assert cs.level == level
+
+        # a compression level is not specified (default)
+        elif level == "":
+            cs = CompressionSpec(c_name).compressor
+            assert isinstance(cs, c_type)
+            if c_type in (ZLIB, LZMA):
+                assert cs.level == 6
+            elif c_type == ZSTD:
+                assert cs.level == 3
+
+        # a compression level is invalid
+        elif level.endswith("invalid"):
+            invalid_spec = f"{c_name},{level}" if c_name else level
+            with pytest.raises(argparse.ArgumentTypeError):
+                CompressionSpec(invalid_spec)
```

Comment:
hmm, split this into 3 separate tests: for defaults, for given levels, for invalid levels and get rid of the if/elif?

Reply:
with separate tests, some of the information in the comments could be also transported via a good test name.

Reply:
Absolutely 👍